engine/tx_sync_task.go

```diff
@@ -0,0 +1,135 @@
+package engine
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/bitcoin-sv/spv-wallet/engine/chainstate"
+	"github.com/bitcoin-sv/spv-wallet/engine/spverrors"
+	"github.com/rs/zerolog"
+)
+
+// delayForBroadcastedTx indicates the time after which a broadcasted transaction should be checked
+// most probably ARC callback hasn't been received in this time so we need to check the transaction status "manually"
+func delayForBroadcastedTx() time.Time {
+	return time.Now().Add(-time.Hour)
+}
+
+// delayForNotBroadcastedTx indicates the time after which a non-broadcasted transaction should be checked.
+// In this case, we don't have to wait for an ARC callback (because it will never come).
+// We're checking the transaction status after potentially enough time has passed for it to be mined.
+func delayForNotBroadcastedTx() time.Time {
+	return time.Now().Add(-10 * time.Minute)
+}
+
+// problematicTxDelay indicates the time after which a transaction with an unknown status will be marked as problematic
+// This is to prevent the system from trying to check old transactions that are not likely to be valid anymore
+// NOTE: The SYNC task will check such "old" transactions at least once before marking them as problematic
+func problematicTxDelay() time.Time {
+	return time.Now().Add(-24 * time.Hour)
+}
+
+// processSyncTransactions is a crucial periodic task which try to query transactions which cannot be considered as finalized
+// 1. It gets transaction IDs to sync
+// 2. For every transaction check the status using chainstate.QueryTransaction
+// 3. If found - change the status
+// 4. On error - try to rebroadcast (if needed) or
+func processSyncTransactions(ctx context.Context, client *Client) {
+	logger := client.Logger()
+	db := client.Datastore().DB()
+	chainstateService := client.Chainstate()
+
+	recoverAndLog(logger)
+
+	var txIDsToSync []struct {
+		ID string
+	}
+	queryIDsCtx, cancel := context.WithTimeout(ctx, 60*time.Second)
+	defer cancel()
+	err := db.
+		WithContext(queryIDsCtx).
+		Model(&Transaction{}).
+		Where("tx_status = ? AND created_at < ?", TxStatusBroadcasted, delayForBroadcastedTx()).
+		Or("tx_status = ? AND created_at < ?", TxStatusCreated, delayForNotBroadcastedTx()).
+		Or("tx_status IS NULL"). // backward compatibility
+		Find(&txIDsToSync).Error
+	if err != nil {
+		logger.Error().Err(err).Msg("Cannot fetch transactions to sync")
+		return
+	}
+
+	logger.Info().Msgf("Transactions to SYNC: %d", len(txIDsToSync))
+
+	for _, record := range txIDsToSync {
+		txID := record.ID
+		tx, err := getTransactionByID(ctx, "", txID, WithClient(client))
+		if tx == nil || err != nil {
+			logger.Error().Msg("Cannot get transaction by ID even though the ID was returned from DB")
+			continue
+		}
+		saveTx := func() {
+			if err := tx.Save(ctx); err != nil {
+				logger.Error().Err(err).Str("txID", txID).Msg("Cannot update transaction")
+			}
```

Comment:
i can see lots of code like this 

```go
if err := tx.Save(ctx); err != nil {
				logger.Error().Err(err).Str("txID", txID).Msg("Cannot update transaction")
			}
```

why not create function and use it for save tx 
```go
func saveTransaction(ctx context.Context, tx *Transaction, logger *zerolog.Logger) {
    if err := tx.Save(ctx); err != nil {
        logger.Error().Err(err).Str("txID", tx.ID).Msg("Cannot update transaction")
    }
}
```

Reply:
I can see it too - but I'd like to minimize the refactorization of the code in this PR - because I wanted to focus on what we  need to do for now. 

However, When I'll implement the transaction repository (which it's most probably one of the next task) I would consider your suggestion.